content/blog/ebpf-http2-tracing/index.md

```diff
@@ -0,0 +1,145 @@
+
+## When does Wireshark fail to decode HTTP/2 headers?
+
+[Wireshark](https://www.wireshark.org/) is a well-known network sniffing tool that can capture HTTP/2. However, Wireshark sometimes fails to decode the HTTP/2 headers. Let’s see this in action.
+
+If we launch Wireshark _before_ we start our gRPC demo application, we see captured HTTP/2 messages in Wireshark:
+
+::: div image-l
+<svg title='Wireshark captured HTTP/2 HEADERS frame.' src='wireshark-http2.png' />
+:::
+
+Let’s focus on the [HEADERS frame](https://datatracker.ietf.org/doc/html/rfc7540#section-6.2), which is equivalent to the headers in HTTP 1.x, and records metadata about the HTTP/2 session. We can see that one particular HTTP/2 header block fragment has the raw bytes `bfbe`. In this case, the raw bytes encode the `grpc-status` and `grpc-message` headers. These are decoded correctly by Wireshark as follows:
+
+::: div image-l
+<svg title='Wireshark is able to decode HTTP/2 HEADERS if launched before the message stream starts.' src='wireshark-http2-headers-captured.png' />
+:::
+
+Next, let’s launch Wireshark _after_ launching gRPC client & server. The same messages are captured, but the raw bytes can no longer be decoded by Wireshark:
+
+::: div image-l
+<svg title='Wireshark cannot decode HTTP/2 HEADERS if launched after the message stream starts.' src='wireshark-http2-headers-not-captured.png' />
+:::
+
+Here, we can see that the `Header Block Fragment` still shows the same raw bytes, but the clear-text headers cannot be decoded.
+
+To replicate the experiment for yourself, follow the directions [here](https://github.com/pixie-io/pixie-demos/tree/main/http2-tracing).
+
+## HPACK: the bane of the Wireshark
+
+Why can’t Wireshark decode HTTP/2 headers if it is launched after our gRPC application starts transmitting messages?
+
+It turns out that HTTP/2 uses [HPACK](https://httpwg.org/specs/rfc7541.html) to encode & decoder headers, which compresses the headers and [greatly improves the efficiency over HTTP 1.x](https://blog.cloudflare.com/hpack-the-silent-killer-feature-of-http-2/).
+
+HPACK works by maintaining identical lookup tables at the server and client. Headers and/or their values are replaced with their indices in these lookup tables. Because most of the headers are repetitively transmitted, they are replaced by indices that use much less bytes than clear-text headers. HPACK therefore uses significantly less network bandwidth. This effect is amplified by the fact that multiple HTTP/2 sessions can multiplex over the same connection.
+
+The figure below illustrates the table maintained by the client and server for response headers. New header name and value pairs are appended into the table, displacing the old entries if the size of the lookup tables reaches its limit. When encoding, the clear text headers are replaced by their indices in the table. For more info, take a look at [the official RFC](http://http2.github.io/compression-spec/compression-spec.html).
```

Comment:
Replace `http://http2.github.io/compression-spec/compression-spec.html` with `https://httpwg.org/specs/rfc7541.html`. Such that they are consistent with the previous link of HPACK.